### PR TITLE
[region-isolation] Initial refactoring work for supporting Transferring Args and Results

### PR DESCRIPTION
**Pull request by gottesmm**
Just in progress.

### CODE REVIEW

**Comment by gottesmm** [comment]
Just beginning to test out some support/refactoring that I am doing here before adding in the larger part of work here.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -4260,7 +4309,7 @@ class SILParameterInfo {
   bool operator==(SILParameterInfo rhs) const {
     return getInterfaceType() == rhs.getInterfaceType() &&
            getConvention() == rhs.getConvention() &&
-           getDifferentiability() == rhs.getDifferentiability();
+           getOptions().toRaw() == rhs.getOptions().toRaw();
```
Comment (xedin) [inline include/swift/AST/Types.h:4312]: I think in all the spots where `==` + `.toRaw()` it used you actually want to use `.containsOnly`.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -4309,64 +4358,114 @@ inline bool isIndirectFormalResult(ResultConvention convention) {
...
-};
-
 /// A result type and the rules for returning it.
 class SILResultInfo {
-  CanType Type;
-  ResultConvention Convention;
-  SILResultDifferentiability Differentiability;
+public:
+  enum Flag : uint8_t {
+    None = 0x0,
```
Comment (xedin) [inline include/swift/AST/Types.h:4365]: I'm not sure whether `None` should be a `Flag`. Seems like empty `Options` should act the role.

File: include/swift/Basic/OptionSet.h [hunk 3]
```diff
@@ -140,6 +140,12 @@ class OptionSet {
     return lhs;
   }
 
+  constexpr bool operator==(OptionSet lhs) const {
```
Comment (xedin) [inline include/swift/Basic/OptionSet.h:143]: There is a comment about regarding why `==` and `!=` are not declared.

File: include/swift/AST/Types.h (hunk 1 above)
Comment (gottesmm) [reply include/swift/AST/Types.h:4312]: No. I am trying to compare against all possible flags.

Comment (gottesmm) [reply include/swift/AST/Types.h:4312]: I misunderstood the API... you are correct.

**Comment by gottesmm** [comment]
Actual swift-syntax PR: https://github.com/apple/swift-syntax/pull/2411

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/2411

@swift-ci smoke test

File: include/swift/Demangling/TypeDecoder.h [hunk 4]
```diff
@@ -106,22 +107,31 @@ enum class ImplParameterConvention {
   Pack_Inout,
 };
 
-enum class ImplParameterDifferentiability {
-  DifferentiableOrNotApplicable,
-  NotDifferentiable
+enum class ImplParameterInfoFlags : uint8_t {
+  None = 0x0,
```
Comment (xedin) [inline include/swift/Demangling/TypeDecoder.h:111]: `None` is still around here. And the enum should be singular named?

File: include/swift/Demangling/TypeDecoder.h [hunk 5]
```diff
@@ -106,22 +107,31 @@ enum class ImplParameterConvention {
...
+  NotDifferentiable = 0x1,
+};
+
+using ImplParameterInfoOptions = OptionSet<ImplParameterInfoFlags>;
+
+enum class ImplResultInfoFlags : uint8_t {
+  None = 0x0,
```
Comment (xedin) [inline include/swift/Demangling/TypeDecoder.h:118]: `None` is still around here. And the enum should be singular named?

File: lib/Serialization/ModuleFormat.h [hunk 6]
```diff
@@ -395,6 +395,16 @@ enum class SILParameterDifferentiability : uint8_t {
   NotDifferentiable,
 };
 
+/// These IDs must \em not be renumbered or reordered without incrementing the
+/// module version.
+enum class SILParameterInfoFlags : uint8_t {
+  None = 0x0,
```
Comment (xedin) [inline lib/Serialization/ModuleFormat.h:401]: `None` is still around here. And the enum should be singular named?

File: lib/Serialization/ModuleFormat.h [hunk 7]
```diff
@@ -407,13 +417,16 @@ enum class ResultConvention : uint8_t {
...
 
-// These IDs must \em not be renumbered or reordered without incrementing
-// the module version.
-enum class SILResultDifferentiability : uint8_t {
-  DifferentiableOrNotApplicable = 0,
-  NotDifferentiable,
+/// These IDs must \em not be renumbered or reordered without incrementing the
+/// module version.
+enum class SILResultInfoFlags : uint8_t {
+  None = 0x0,
```
Comment (xedin) [inline lib/Serialization/ModuleFormat.h:423]: `None` is still around here. And the enum should be singular named?

**Review by xedin — APPROVED** [review]

**Comment by gottesmm** [comment]
https://github.com/apple/swift-syntax/pull/2411

@swift-ci smoke test

**Pull request merged**